### ISSUE TITLE
Showing dialog for exception: addCharm() takes 3 positional arguments but 4 were given

### ISSUE THREAD

**Issue by jeremybrown82** [conjure-up/conjure-up#1628]
## Report

Thank you for trying **conjure-up**! Before reporting a bug please make sure you've gone through this checklist:

- [ ] Is this problem already documented at https://docs.ubuntu.com/conjure-up/en/troubleshoot#common-spell-problems no
- [ ] Is conjure-up running inside a virtual machine? no
- [ ] Is this problem reproducible with `sudo snap refresh conjure-up --edge`? yes

## Please provide the output of the following commands

```
which juju
/snap/bin/juju

juju version
2.7-beta1-centos7-amd64

which conjure-up
/snap/bin/conjure-up

conjure-up --version
conjure-up 2.6.8


which lxc
/snap/bin/lxc

/snap/bin/lxc config show
config:
  core.https_address: '[::]'

/snap/bin/lxc version
Client version: 3.16
Server version: 3.16

cat /etc/lsb-release
cat: /etc/lsb-release: No such file or directory

```

Please attach tarball of **~/.cache/conjure-up**:

```
tar cvzf conjure-up.tar.gz ~/.cache/conjure-up
```

## Sosreport

Please attach a sosreport:

```
sudo apt install sosreport
sosreport
```

The resulting output file can be attached to this issue.

### MAAS Users
Which version of MAAS?

## Commands ran

Please outline what commands were run to install and execute conjure-up:
conjure-up
## Additional Information

2019-09-04 07:59:03,905 [INFO] conjure-up/openstack-novalxd - common.py:19 - Deploying Applications.
2019-09-04 07:59:04,030 [DEBUG] conjure-up/openstack-novalxd - events.py:52 - Setting Error at conjureup/events.py:149
2019-09-04 07:59:04,031 [ERROR] conjure-up/openstack-novalxd - events.py:161 - Unhandled exception in <Task finished coro=<do_deploy() done, defined at /snap/conjure-up/1055/lib/python3.6/site-packages/conjureup/controllers/juju/deploy/common.py:11> exception=TypeError('addCharm() takes 3 positional arguments but 4 were given',)>
Traceback (most recent call last):
  File "/snap/conjure-up/1055/lib/python3.6/site-packages/conjureup/controllers/juju/deploy/common.py", line 30, in do_deploy
    await app.juju.client.deploy(fn)
  File "/snap/conjure-up/1055/lib/python3.6/site-packages/juju/model.py", line 1308, in deploy
    await handler.execute_plan()
  File "/snap/conjure-up/1055/lib/python3.6/site-packages/juju/model.py", line 1991, in execute_plan
    result = await method(*step.args)
TypeError: addCharm() takes 3 positional arguments but 4 were given
2019-09-04 07:59:04,059 [DEBUG] conjure-up/openstack-novalxd - __init__.py:27 - Showing dialog for exception: addCharm() takes 3 positional arguments but 4 were given
2019-09-04 07:59:37,957 [DEBUG] conjure-up/openstack-novalxd - events.py:52 - Setting Shutdown at conjureup/events.py:145

**Comment by jeremybrown82** [comment]
[https://gofile.io/?c=RtOZvG](url)

tar cvzf conjure-up.tar.gz ~/.cache/conjure-up

**Comment by 7ko7** [comment]
It's not a conjure-up problem. I solved this by using the candidate version of juju which is currently: 2.6.9+2.6-f190cd8.

juju refresh --channel=2.6/candidate juju

**Comment by AurelienLourot** [comment]
**TL;DR** I'm preparing a pull request to fix this.

----------

@7ko7 you probably mean

```
sudo snap refresh --channel=2.6/candidate juju
```

This didn't help for me and actually I wonder how this can have helped in your case because your command changes the version of juju installed on your system but it doesn't affect the version of juju installed by the conjure-up snap.

Interestingly we now see this failure on Travis since the last commit 20 days ago:
* https://github.com/conjure-up/conjure-up/commit/11f26e374a74b2eb8337622b00e025e9afe347dd
* https://travis-ci.org/conjure-up/conjure-up/jobs/618355355

In this commit we moved from juju 2.6.9 to 2.6.10.

Also interesting to mention that the exception is happening in

```
  File "/snap/conjure-up/1058/lib/python3.6/site-packages/juju/model.py", line 1991, in execute_plan
    result = await method(*step.args)
TypeError: addCharm() takes 3 positional arguments but 4 were given
```

This part of the code got refactored recently:
* https://github.com/juju/python-libjuju/pull/332
* https://github.com/juju/python-libjuju/blob/master/docs/changelog.rst

This refactoring seems to solve the issue. If we pin the latest `python-libjuju` here in conjure-up the problem disappears. Here is the fix:
https://github.com/AurelienLourot/conjure-up/commit/c0095debf0d4807e22182ba0065a713956e1ffc9

If you want you can build and use that special version locally already now:

```
sudo snap remove conjure-up
sudo snap install snapcraft --classic
git clone https://github.com/AurelienLourot/conjure-up
cd conjure-up/
snapcraft
sudo snap install conjure-up_2.6.10-*_amd64.snap --dangerous --classic
```

**Comment by AurelienLourot** [comment]
Duplicate bug: https://bugs.launchpad.net/conjure-up/+bug/1842321

**Comment by AdamIsrael** [comment]
I'm afraid the fix doesn't work for me. 
It gets further, but dies with:

```
entity() got an unexpected keyword argument 'include_stats'
```

[conjure-up.log](https://github.com/conjure-up/conjure-up/files/3907789/conjure-up.log)

**Comment by AurelienLourot** [comment]
Too bad, it worked for me but I was deploying OpenStack and it looks like you're deploying Kubernetes. Can you open a new bug with some steps to reproduce? Thanks!

**Comment by dbwest** [comment]
steps to reproduce are using maas try and deploy charmed kubernetes. can use all the defaults. this used to work. it no longer works.
